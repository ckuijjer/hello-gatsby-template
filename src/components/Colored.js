import React from 'react';
import colors from '../colors';

const shadesOfOrange = [
  colors.nnLightOrange,
  colors.nnMediumOrange,
  colors.nnDarkOrange,
  colors.nnMediumOrange,
];

// x => false => [] => []
// x => x == 2 => [1, 2, 3] => [[1,2],[3]]
// x => x == 3 => [1, 2, 3] => [[1,2,3]]
const splitAfter = (fn, list) => {
  const result = list.reduce(
    (acc, cur, i) => {
      acc[acc.length - 1].push(cur);

      if (fn(cur, i)) {
        acc.push([]);
      }

      return acc;
    },
    [[]],
  );

  if (result[result.length - 1].length === 0) {
    result.pop();
  }

  return result;
};

export const getWordsPerColor = x => (x > 12 ? 3 : x > 5 ? 2 : 1);

const Colored = ({ children = '' }) => {
  let words = children.split(' ');
  if (words.length === 1 && words[0] === '') words = []; // no text, then make sure it'll return null

  const wordsPerColor = getWordsPerColor(words.length);

  const sentences = splitAfter(x => x.endsWith('.'), words);

  const result = sentences
    .flatMap(sentence =>
      splitAfter((x, i) => (i + 1) % wordsPerColor === 0, sentence).map(
        (group, i) => (
          <span style={{ color: shadesOfOrange[i % shadesOfOrange.length] }}>
            {group.join(' ')}
          </span>
        ),
      ),
    )
    .map((value, i, arr) =>
      i + 1 === arr.length ? <>{value}</> : <>{value} </>,
    ); // add a space between the words

  return result;
};

export default Colored;
