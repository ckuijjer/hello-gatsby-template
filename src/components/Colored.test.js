import React from 'react';
import renderer from 'react-test-renderer';

import Colored, { getWordsPerColor } from './Colored';

describe('getWordsPerColor', () => {
  it('1 word 1 color', () => {
    expect(getWordsPerColor(1)).toBe(1);
  });

  it('5 word 1 color', () => {
    expect(getWordsPerColor(5)).toBe(1);
  });

  it('6 word 2 color', () => {
    expect(getWordsPerColor(6)).toBe(2);
  });

  it('12 word 2 color', () => {
    expect(getWordsPerColor(12)).toBe(2);
  });

  it('13 word 3 color', () => {
    expect(getWordsPerColor(13)).toBe(3);
  });
});

describe('Colored', () => {
  it('renders empty text correctly', () => {
    const tree = renderer.create(<Colored />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders a single word correctly', () => {
    const tree = renderer.create(<Colored>one</Colored>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders five words correctly', () => {
    const tree = renderer
      .create(<Colored>one two three four five</Colored>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders six words correctly', () => {
    const tree = renderer
      .create(<Colored>one two three four five six</Colored>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders twelve words correctly', () => {
    const tree = renderer
      .create(
        <Colored>
          one two three four five six seven eight nine ten eleven twelve
        </Colored>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders thirteen words correctly', () => {
    const tree = renderer
      .create(
        <Colored>
          one two three four five six seven eight nine ten eleven twelve
          thirteen
        </Colored>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders two sentences correctly', () => {
    const tree = renderer
      .create(<Colored>sentence one. sentence two</Colored>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
