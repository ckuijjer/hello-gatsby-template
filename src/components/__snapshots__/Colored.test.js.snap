// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colored renders a single word correctly 1`] = `
<span
  style={
    Object {
      "color": "#ee7f00",
    }
  }
>
  one
</span>
`;

exports[`Colored renders empty text correctly 1`] = `null`;

exports[`Colored renders five words correctly 1`] = `
Array [
  <span
    style={
      Object {
        "color": "#ee7f00",
      }
    }
  >
    one
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    two
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#e64415",
      }
    }
  >
    three
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    four
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ee7f00",
      }
    }
  >
    five
  </span>,
]
`;

exports[`Colored renders six words correctly 1`] = `
Array [
  <span
    style={
      Object {
        "color": "#ee7f00",
      }
    }
  >
    one two
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    three four
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#e64415",
      }
    }
  >
    five six
  </span>,
]
`;

exports[`Colored renders thirteen words correctly 1`] = `
Array [
  <span
    style={
      Object {
        "color": "#ee7f00",
      }
    }
  >
    one two three
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    four five six
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#e64415",
      }
    }
  >
    seven eight nine
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    ten eleven twelve
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ee7f00",
      }
    }
  >
    thirteen
  </span>,
]
`;

exports[`Colored renders twelve words correctly 1`] = `
Array [
  <span
    style={
      Object {
        "color": "#ee7f00",
      }
    }
  >
    one two
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    three four
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#e64415",
      }
    }
  >
    five six
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    seven eight
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ee7f00",
      }
    }
  >
    nine ten
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    eleven twelve
  </span>,
]
`;

exports[`Colored renders two sentences correctly 1`] = `
Array [
  <span
    style={
      Object {
        "color": "#ee7f00",
      }
    }
  >
    sentence
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    one.
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ee7f00",
      }
    }
  >
    sentence
  </span>,
  " ",
  <span
    style={
      Object {
        "color": "#ea650d",
      }
    }
  >
    two
  </span>,
]
`;
